features/support/form_helpers.rb

```diff
@@ -200,16 +212,184 @@ def fill_form(locator = nil, options = {})
+  end
+
+  def get_answers_for_validated_questions
+    options = {}
+
+    options.merge!(pass_text_validation)
+    options.merge!(pass_pricing_validation)
+
+    return :gotosave if pass_document_upload_validation
+    return :gotosave if pass_checked_boxes_upper_limit
+
+    # fields with custom validation:
+    options.merge!(pass_vat_number_validation)
+    options.merge!(pass_companies_house_number_validation)
+  end
+
+  def choose_options_for_select_fields
+    selects = find_elements_by_xpath("//select")
+    if selects.length > 0
+      selects.each do |element|
+        field = all_fields(element["name"])[0]
+        field_options = find_elements_by_xpath("//select[@name='#{element['name']}']/descendant::option")
+        selected = field_options[rand(field_options.count - 1)]
+        selected.select_option
+        puts(element["name"] + "=> " + selected["value"])
+      end
+    end
+  end
+
+  def open_categories(categories_headings)
+    categories_headings.each do |category_heading|
+      category_heading.click
+    end
+  end
+
+  def find_and_click_submit_button
+    submit_button = find_elements_by_xpath("//input[@class='button-save']")[0].value
+    if submit_button == 'Save and continue'
+      click_on 'Save and continue'
+      false
+    else
+      click_on submit_button
+      true
+    end
+  end
+
+  def answer_all_service_questions(prompt_text)
+    while find_elements_by_xpath("//a[text()='#{prompt_text}']").length > 0
+      if find_elements_by_xpath("//input[@value='Mark as complete']").length > 0
+        return
+      end
+      next_answer_question_link = page.all(:xpath, "//a[text()='#{prompt_text}']")[0]
+      next_answer_question_link.click
+      # To debug particular section, use below line instead of above lines:
+      # find_elements_by_xpath("//span[text()='Section title here']/following::a[text()='Answer question']")[0].click
+      answer_service_section
+    end
+  end
+
+  def answer_service_section
+    is_end_of_section = false
+    fail_counter = 0
+    while is_end_of_section == false
+      categories_headings = find_elements_by_xpath("//h3[@class='categories-heading']")
+      if is_there_validation_header?
+        fail_counter += 1
+        if fail_counter > 7
+          abort("Too many failures")
+        end
+        options = get_answers_for_validated_questions
+        unless options == :gotosave
+          answer = fill_form with: options
+        end
+      elsif categories_headings.length > 0
+        open_categories(categories_headings)
+        check_first_checkbox
+        options = :gotosave
+      else
+        answer = fill_form
+      end
+      unless options && options == :gotosave
+        choose_options_for_select_fields
+        merge_fields_and_print_answers(answer)
+      end
+      page.save_screenshot("screenshot_#{Time.new}.png")
+      is_end_of_section = find_and_click_submit_button
+      is_end_of_section = false if is_there_validation_header?
+    end
+  end
```

Comment:
This might be cool to explore recursion. Might not be possible depending on what's being done but we can have a go :)

```
def answer_service_section(end_of_section = false, fail_counter = 0)
  break if end_of_section

  categories_headings = find_elements_by_xpath("//h3[@class='categories-heading']")
  if is_there_validation_header?
    fail_counter +1
    if fail_counter > 7
      abort("Too many failures")
    end
    options = get_answers_for_validated_questions
    unless options == :gotosave
      answer = fill_form with: options
    end
  elsif categories_headings.length > 0
    open_categories(categories_headings)
    check_first_checkbox
    options = :gotosave
  else
    answer = fill_form
  end
  unless options && options == :gotosave
    choose_options_for_select_fields
    merge_fields_and_print_answers(answer)
  end
  page.save_screenshot("screenshot_#{Time.new}.png")
  
  end_of_section = find_and_click_submit_button || is_there_validation_header?

  answer_service_section(end_of_section, fail_counter)
end
```

Reply:
Tempting, what would recursion give us here though?